### DISCUSSION TITLE
after behaves different on dev and production

### DISCUSSION THREAD

**Discussion by alexmilde** [tailwindlabs/tailwindcss#10837] [Help]
**What version of Tailwind CSS are you using?**

3.2.7

**What build tool (or framework if it abstracts the build tool) are you using?**

"laravel-mix": "^6.0.39"

**What version of Node.js are you using?**

18.14.2

**What browser are you using?**

Chrome, Safarai, Firefox

**What operating system are you using?**

macOS

**Reproduction URL**

[https://play.tailwindcss.com/sqZIMAtsFw](https://play.tailwindcss.com/sqZIMAtsFw)

As it only occurs when built in production mode it looks fine on the link, butt will be broken when built as prod

**Describe your issue**

The "after-black" class is missing in production, if it wasnt set without "after" before.
Building in develop mode works correct

<img width="322" alt="Bildschirm­foto 2023-03-21 um 14 49 26" src="https://user-images.githubusercontent.com/2182118/226628195-3bbe6883-2e52-4cf3-abc0-365cc64035ad.png">

<img width="321" alt="Bildschirm­foto 2023-03-21 um 14 49 51" src="https://user-images.githubusercontent.com/2182118/226628260-0ac1e067-aab2-4ff6-aa7e-340e870de769.png">

**Comment by alexmilde** [comment]
thanks for tailwind! its amazing :)

**Comment by thecrypticace** [comment]
@alexmilde Tailwind CSS itself doesn't conditionally produce different output in dev vs production. This likely has to do with the build process being used. I set up an installation of Laravel + Laravel Mix with the repro and a production build and it doesn't showcase the bug: https://github.com/thecrypticace/tailwind-mix-after-repro

Could you provide a reproduction by chance?

**Reply by alexmilde** [reply]
Thanks for the quick reply. 
I will try to narrow it down and provide a reproduction.

**Comment by alexmilde** [comment]
I could pin it down to the following:

```
<ul>
    <li class="after:after:font-black">This creates the error </li>
    <li class="border after:content-['Test2'] font-black after:font-black">Hello: </li>
    <li class="after:content-['Test2'] after:font-black">Hello: </li>
</ul>
```

using a wrong duplicated after:after will result in different outputs when compiling as dev vs compiling as prod.

i could reproduce this in your repo and in my system aswell.

im not shure whether this is a laravel-mix issue or a tailwind issue though.

**Reply by thecrypticace** [reply]
This is because `::after::after` isn't a valid CSS selector (at least — it is not today) and CSS Nano is collapsing the declarations of both `after:after:font-black` and `after:font-black` into the same rule list:

```css
.after\:after\:font-black::after::after, .after\:font-black::after {
  content: var(--tw-content);
  font-weight: 900;
}
```

The browser sees this and promptly throws out the whole rule. So, if you look, it's likely present in your CSS. It's just the browser is throwing it out.

We still generate CSS for `after:after:font-black` because we don't special case the presence of multiple variants. Stacking multiple of the same variant is valid (though generally unnecessary):

<img width="402" alt="Screen Shot 2023-03-21 at 16 30 32" src="https://user-images.githubusercontent.com/614993/226733756-66a5ce56-1ba0-43d7-9d66-b158a52a547b.png">
<img width="422" alt="Screen Shot 2023-03-21 at 16 30 49" src="https://user-images.githubusercontent.com/614993/226733775-ef0e3d33-a693-4741-93e9-72dbe7496ca6.png">
<img width="598" alt="Screen Shot 2023-03-21 at 16 46 19" src="https://user-images.githubusercontent.com/614993/226736777-0f21a7db-e29f-44a9-80ab-754d090f6519.png">


So we're gonna say not-a-bug from the perspective of Tailwind CSS — it would be best to not write `after:after:` in a utility since it's not valid today _and_ will cause bugs like this to pop up in minifiers.

**Reply by alexmilde** [reply]
Thought its a minifiing issue. Thanks.
I somehow assumed tailwind would only generate valid css and throw it out, or at least warn about it.

Of course it would be best not to write invalid selectors.
But sometimes a copy paste error slips through review.

Thanks again.
